packages/core/src/actions.ts

```diff
@@ -713,7 +718,7 @@ export function resolveActions<TContext, TEvent extends EventObject>(
             if (exec && preservedContexts) {
               const contextIndex = preservedContexts.length - 1;
               resolvedActionObject.exec = (_ctx, ...args) => {
```

Comment:
This is a dangerous mutation - it kinda works OK now (I wasn't able to find a compelling failing test case) but I would call this to be rather coincidental. 

However, in combination with this PR it leads to a memory leak and a broken behavior. I can't showcase easily the memory leak but basically, a function action gets converted to an `{ exec }` object, and then it gets mutated here (each time). So a new `exec` always wraps the previous which wraps the previous one which wraps the previous one and so on.

I was able to write down a failing test for the broken behavior though: https://github.com/statelyai/xstate/commit/c49e058e094c4d565b95668a6cbad5b609158d78 . To fix this you should follow the advice given here: https://github.com/statelyai/xstate/pull/2430/files#r675457307 . Just do not resolve action objects in the constructor and in the `formatTransition`.

This is quite a foot gun and how the whole action resolving works is a little bit convoluted for what it is. I think this is a place where the codebase could be refactored and cleaned up in the process - it's rather a v5 item though.